[Loh Xian Ze, Bryan] iP

Anonymxtrix opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Anonymxtrix committed:
Implement skeletal version of Duke (#2)

* Add greet, echo, exit functions
* Refactor console related logic in Duke class

Anonymxtrix committed:
Add ability to store and display text (#4)

* Add skeleton for storing and displaying text

Anonymxtrix committed:
Add ability to mark tasks as done (#6)

* Add skeleton for ability to mark tasks as done
* Add Response and TaskCollection classes
* Add add message method and additional constructors to Response

Anonymxtrix committed:
Add support for tracking todos deadlines and events (#8)

* Add Deadline Event and ToDo classes
* Add create todo, deadline and event commands
* Add Action interface and AddTask, CompleteTask and ListTask classes
* Add WelcomeUser and GoodbyeUser actions
* Add actions queue to Duke class
* Add Command enum and Request classes
* Modify default request case to add ToDo task
* Add TaskCollectionRequest abstract class
* Fix wrong comments and add missing lines

Anonymxtrix committed:
Add semi-automatic tests using input/output redirection (#10)

* Add success tests for list, todo, deadline, event, done, bye

Anonymxtrix committed:
Add error handling to Duke application (#12)

* Add UserException class and throw respective exceptions in the Request classes
* Add UserExceptions thrown by Command and Requests to tests

Anonymxtrix committed:
Add support for deleting tasks from the list (#14)

* Add support for deleting tasks from the TaskCollection
* Add tests for delete and UserException thrown by DeleteRequest

Anonymxtrix committed:
Add date and time support to Deadline and Event classes

Anonymxtrix committed:
Add support for saving tasks in hard disk (#16)

* Add Storage class to read from and write to text file
* Add TaskType enum
* Add ability to save Tasks to Storage
* Fix missing end of file line in .gitignore

Anonymxtrix committed:
Add Ui class to deal with user interactions (#20)

* Add Ui class to deal with user interactions

Anonymxtrix committed:
Move files into individual packages (#22)

* Move files into individual packages

Anonymxtrix committed:
Add JUnit tests to test the behavior of the code (#24)

* Add some tests for Response and Task classes

Anonymxtrix committed:
Refactor code to comply with CS2103T coding standard

Anonymxtrix committed:
Add ability to find Tasks by regex String

Anonymxtrix committed:
Merge branch 'branch-Level-7' into new-master

Anonymxtrix committed:
Merge branch 'branch-Level-8' into new-master

* Add date and time support to Deadline and Event classes

Anonymxtrix committed:
Add JavaDoc to Duke class header

Anonymxtrix committed:
Merge branch 'branch-A-JavaDoc' into new-master

Anonymxtrix committed:
Merge branch 'branch-A-CodingStandard' into new-master

Anonymxtrix committed:
Merge tag 'branch-Level-9' into new-master

* Add ability to find Tasks by regex String

Anonymxtrix committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Anonymxtrix committed:
Modify gradle main class name

Anonymxtrix committed:
Merge branch 'branch-A-Gradle'

Anonymxtrix committed:
Add checkstyle.xml and suppressions.xml to project

Anonymxtrix committed:
Fix all files to follow checkstyle rules

Anonymxtrix committed:
Merge branch 'branch-A-CheckStyle'

wengYing227 commented on src/main/java/duke/task/ToDo.java:
```diff
@@ -0,0 +1,35 @@
+     * @return The String representation of a duke.task.ToDo.
+     */
+    @Override
+    public String toString() {
```
I think toString comment is not compulsory

wengYing227 commented on src/test/java/duke/ResponseTest.java:
```diff
@@ -0,0 +1,18 @@
+                ResponseTest.LINE + System.lineSeparator()
+                + ResponseTest.MESSAGE_INDENT + "message" + System.lineSeparator()
+                + ResponseTest.LINE
+                , new Response("message").toString());
```
Comma should be at the end of line 15

wengYing227 commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * Parse the input String into an equivalent duke.task.TaskType.
```
It's better if you use the word _Parses_ instead of _Parse_.

Anonymxtrix replied:
Nice catch. Thanks.

wengYing227 commented on src/main/java/duke/task/TaskCollection.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * Get the number of Tasks in the duke.task.TaskCollection.
```
You should use the third person singular verb for _Get_

Anonymxtrix replied:
What is "third person singular verb"? O.o

wengYing227 commented on src/main/java/duke/task/TaskCollection.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * Get the number of Tasks in the duke.task.TaskCollection.
```
Same as below

wengYing227 reviewed (commented):
LGTM. :thumbsup::thumbsup::thumbsup: You've done a really good job in code styling!!

itsraveen commented on src/main/java/duke/Response.java:
```diff
@@ -0,0 +1,68 @@
+package duke;
+
+/**
+ * A duke.Response from the application to some user input.
```
I think in your java docs, it might be easier to read and understand if you do not include your package in every the JavaDocs for every method.  Because it is clear from line 1 that this is part of the duke package. But the content in the JavaDocs are fine.

itsraveen commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,27 @@
+     * Prints the duke.Response to the console.
+     * @param response The duke.Response to be printed.
+     */
+    public void printResponse(Response response) {
+        System.out.println(response.toString());
+    }
```
Is this method necessary? I think it would be okay to directly use the System.out.println method instead of having this method.

Anonymxtrix replied:
Yes it is, to respect abstraction.

itsraveen commented on src/main/java/duke/request/Command.java:
```diff
@@ -0,0 +1,47 @@
+ * A duke.request.Command represents a string in a duke.request.Request that instructs the application to
+ * perform an Action.
+ */
+public enum Command {
+    BYE("bye"),
+    FIND("find"),
+    LIST("list"),
+    DONE("done"),
+    DELETE("delete"),
+    DEADLINE("deadline"),
+    EVENT("event"),
+    TODO("todo");
+
```
Great that you used enum. Makes the entire code a lot more concise and easier to read. 👍

Anonymxtrix replied:
Thanks!

itsraveen commented on src/main/java/duke/request/Request.java:
```diff
@@ -0,0 +1,58 @@
+        String commandInput = substrings.length > 1 ? substrings[1] : "";
+        Command command = Command.parseFrom(commandString);
+
+        switch (command) {
+        case BYE:
+            return new ByeRequest();
+
+        case DONE:
+            return new DoneRequest(taskCollection, commandInput);
+        case DELETE:
+            return new DeleteRequest(taskCollection, commandInput);
+        case LIST:
+            return new ListRequest(taskCollection);
+        case FIND:
+            return new FindRequest(taskCollection, commandInput);
+
+        case DEADLINE:
+            return new DeadlineRequest(taskCollection, commandInput);
+        case EVENT:
+            return new EventRequest(taskCollection, commandInput);
+        case TODO:
+            return new ToDoRequest(taskCollection, commandInput);
+        default:
+            throw new RuntimeException(String.format(
+                "duke.request.Command %s exists but is not mapped to a corresponding duke.request.Request.",
+                commandString
```
Great use of switch statements. Makes it a lot easier to understand. However, can remove the unnecessary empty lines between cases.

Anonymxtrix replied:
Thanks!

itsraveen commented on src/test/java/duke/ResponseTest.java:
```diff
@@ -0,0 +1,18 @@
+
+public class ResponseTest {
+    private static String LINE = "    ____________________________________________________________";
+    private static String MESSAGE_INDENT = "     ";
```
You can use the \t escape sequence in your string to provide tab spaces. However, if you were planning to customise the amount of tab space then this works too!

itsraveen reviewed (commented):
Code wise mostly great. Easy to understand and done well. Improvements can be made to your JavaDocs. Other than that, looks good to me.

Anonymxtrix committed:
Add a GUI to Duke (#30)

* Add JavaFX to gradle
* Refactor static to instance attributes in Duke class
* Create JavaFX skeleton
* Create a GUI for Duke application
* Create interaction with user
* Add fxml to JavaFX
* Connect MainWindow GUI to Duke application

Anonymxtrix committed:
Merge branch 'branch-Level-10'

Anonymxtrix committed:
Add varargs to Response add method

Anonymxtrix committed:
Merge branch 'branch-A-Varargs'

Anonymxtrix committed:
Add assertions to Duke application.

Currently, all classes in the application do not have assertions that
verify the assumptions made about the program state. As a result, there
are many places that are prone to NullPointerExceptions.

As a result, assertions are added to the actions, Request, Response and
Duke classes so that the runtime can verify these assumptions and flag
them out when breached.

Classes affected:
* actions
  * AddTask
  * CompleteTask
  * DeleteTask
  * FindTasks
  * ListTasks
* Request
* Response
* Duke

Anonymxtrix committed:
Merge pull request #31 from Anonymxtrix/branch-A-Assertions

Add assertions to verify assumptions in core classes

Anonymxtrix committed:
Improve code quality of DialogBox class

Anonymxtrix committed:
Improve code standard of Duke and MainWindow classes

Anonymxtrix committed:
Improve code standard of Main class

Anonymxtrix committed:
Improve code quality in Duke and Response classes

Anonymxtrix committed:
Improve code quality of Storage class

Anonymxtrix committed:
Improve code standard of Ui class

Anonymxtrix committed:
Improve code quality for action package classes

Anonymxtrix committed:
Improve code quality of exception and task package classes

Anonymxtrix committed:
Improve code quality of request and exception package classes

Anonymxtrix committed:
Merge branch 'master' into branch-A-CodeQuality

Anonymxtrix committed:
Merge pull request #32 from Anonymxtrix/branch-A-CodeQuality

Improve code quality of the application

Anonymxtrix committed:
Add more flexibility in search

Anonymxtrix committed:
Merge branch 'branch-C-BetterSearch'

Anonymxtrix committed:
Add user guide to docs/README.md

Anonymxtrix committed:
Merge branch 'branch-A-UserGuide'
